lib/src/idx/ft/analyzer/filter.rs

```diff
@@ -0,0 +1,731 @@
+					Language::Danish => Stemmer::create(Algorithm::Danish),
+					Language::Dutch => Stemmer::create(Algorithm::Dutch),
+					Language::English => Stemmer::create(Algorithm::English),
+					Language::French => Stemmer::create(Algorithm::French),
+					Language::German => Stemmer::create(Algorithm::German),
+					Language::Greek => Stemmer::create(Algorithm::Greek),
+					Language::Hungarian => Stemmer::create(Algorithm::Hungarian),
+					Language::Italian => Stemmer::create(Algorithm::Italian),
+					Language::Norwegian => Stemmer::create(Algorithm::Norwegian),
+					Language::Portuguese => Stemmer::create(Algorithm::Portuguese),
+					Language::Romanian => Stemmer::create(Algorithm::Romanian),
+					Language::Russian => Stemmer::create(Algorithm::Russian),
+					Language::Spanish => Stemmer::create(Algorithm::Spanish),
+					Language::Swedish => Stemmer::create(Algorithm::Swedish),
+					Language::Tamil => Stemmer::create(Algorithm::Tamil),
+					Language::Turkish => Stemmer::create(Algorithm::Turkish),
+				};
+				Filter::Stemmer(a)
+			}
+			SqlFilter::Uppercase => Filter::Uppercase,
+		}
+	}
+}
+
+impl Filter {
+	pub(super) fn from(f: Option<Vec<SqlFilter>>) -> Option<Vec<Filter>> {
+		if let Some(f) = f {
+			let mut r = Vec::with_capacity(f.len());
+			for f in f {
+				r.push(f.into());
+			}
+			Some(r)
+		} else {
+			None
+		}
+	}
+
+	pub(super) fn apply_filters(mut t: Tokens, f: &Option<Vec<Filter>>) -> Tokens {
+		if let Some(f) = f {
+			for f in f {
+				t = t.filter(f);
+			}
+		}
+		t
+	}
+
+	pub(super) fn apply_filter(&self, c: &str) -> FilterResult {
+		match self {
+			Filter::Ascii => Self::deunicode(c),
+			Filter::EdgeNgram(min, max) => Self::edgengram(c, *min, *max),
+			Filter::Lowercase => Self::lowercase(c),
+			Filter::Ngram(min, max) => Self::ngram(c, *min, *max),
+			Filter::Stemmer(s) => Self::stem(s, c),
+			Filter::Uppercase => Self::uppercase(c),
+		}
+	}
+
+	#[inline]
+	fn check_term(c: &str, s: String) -> FilterResult {
+		if s.is_empty() {
+			FilterResult::Ignore
+		} else if s.eq(c) {
+			FilterResult::Term(Term::Unchanged)
+		} else {
+			FilterResult::Term(Term::NewTerm(s))
+		}
+	}
+
+	#[inline]
+	fn lowercase(c: &str) -> FilterResult {
+		Self::check_term(c, c.to_lowercase())
+	}
+
+	#[inline]
+	fn uppercase(c: &str) -> FilterResult {
+		Self::check_term(c, c.to_lowercase())
+	}
+
+	#[inline]
+	fn deunicode(c: &str) -> FilterResult {
+		Self::check_term(c, deunicode(c))
+	}
+
+	#[inline]
+	fn stem(s: &Stemmer, c: &str) -> FilterResult {
+		Self::check_term(c, s.stem(&c.to_lowercase()).into())
+	}
+
+	#[inline]
+	fn ngram(c: &str, min: u16, max: u16) -> FilterResult {
+		let min = min as usize;
+		let l = c.len();
+		if l < min {
+			return FilterResult::Ignore;
+		}
+		let mut ng = vec![];
+		let r1 = 0..(l - min);
+		let max = max as usize;
+		for s in r1 {
+			let mut e = s + max;
+			if e > l {
+				e = l;
+			}
+			let r2 = (s + min)..(e + 1);
+			for p in r2 {
+				let n = &c[s..p];
```

Comment:
It isn't clear how this substring is guaranteed to be aligned to `char` boundaries, so it seems to me like it could panic. Am I missing something?

Reply:
Well spotted!! I am such optimistic :)))